pkg/serviceaccounts/controllers/deleted_dockercfg_secrets.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+	"github.com/golang/glog"
+)
+
+// NumServiceAccountUpdateRetries controls the number of times we will retry on conflict errors.
+// This happens when multiple service account controllers update at the same time.
+const NumServiceAccountUpdateRetries = 10
+
+// DockercfgDeletedControllerOptions contains options for the DockercfgDeletedController
+type DockercfgDeletedControllerOptions struct {
+	// Resync is the time.Duration at which to fully re-list secrets.
+	// If zero, re-list will be delayed as long as possible
+	Resync time.Duration
+}
+
+// NewDockercfgDeletedController returns a new *DockercfgDeletedController.
+func NewDockercfgDeletedController(cl client.Interface, options DockercfgDeletedControllerOptions) *DockercfgDeletedController {
+	e := &DockercfgDeletedController{
+		client: cl,
+	}
+
+	dockercfgSelector := fields.SelectorFromSet(map[string]string{client.SecretType: string(api.SecretTypeDockercfg)})
+	_, e.secretController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.Secrets(api.NamespaceAll).List(labels.Everything(), dockercfgSelector)
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.Secrets(api.NamespaceAll).Watch(labels.Everything(), dockercfgSelector, rv)
+			},
+		},
+		&api.Secret{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{
+			DeleteFunc: e.secretDeleted,
+		},
+	)
+
+	return e
+}
+
+// The DockercfgDeletedController watches for service account dockercfg secrets to be deleted
+// It removes the corresponding token secret and service accoutn references.
+type DockercfgDeletedController struct {
+	stopChan chan struct{}
+
+	client client.Interface
+
+	secretController *framework.Controller
+}
+
+// Runs controller loops and returns immediately
+func (e *DockercfgDeletedController) Run() {
+	if e.stopChan == nil {
+		e.stopChan = make(chan struct{})
+		go e.secretController.Run(e.stopChan)
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (e *DockercfgDeletedController) Stop() {
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+// secretDeleted reacts to a Secret being deleted by looking to see if it's a dockercfg secret for a service account, in which case it
+// it removes the references from the service account and removes the token created to back the dockercfgSecret
+func (e *DockercfgDeletedController) secretDeleted(obj interface{}) {
+	dockercfgSecret := obj.(*api.Secret)
```

Comment:
```
//  * OnDelete will get the final state of the item if it is known, otherwise
//      it will get an object of type cache.DeletedFinalStateUnknown. This can
//      happen if the watch is closed and misses the delete event and we don't
//      notice the deletion until the subsequent re-list.
```